packages/node/src/lib/token-manager.ts

```diff
@@ -0,0 +1,290 @@
+        this.lastError = err
+        timeUntilRefreshInMs = backoff({
+          attempt: this.retryCount,
+          minTimeout: 25,
+          maxTimeout: 1000,
+        })
+        this.pollerTimer = setTimeout(
+          this.pollerLoop.bind(this),
+          timeUntilRefreshInMs
+        ).unref()
+      }
+      let token: AccessToken
+      try {
+        const parsedBody = JSON.parse(body)
+        // TODO: validate JSON
+        token = parsedBody
+        this.tokenEmitter.emit('access_token', { token })
+
+        // Reset our failure count
+        this.retryCount = 0
+
+        // Refresh the token after half the expiry time passes
+        if (token !== undefined && token.expires_in !== undefined) {
+          timeUntilRefreshInMs = Math.floor((token.expires_in / 2) * 1000)
+        } else {
+          timeUntilRefreshInMs = 60 * 1000
+        }
+      } catch (err) {
+        // Something went really wrong with the body, lets surface an error and try again?
+        this.tokenEmitter.emit('access_token', { error: err })
+        this.retryCount = 0
+
+        timeUntilRefreshInMs = backoff({
+          attempt: this.retryCount,
+          minTimeout: 25,
+          maxTimeout: 1000,
+        })
+      }
+    } else if (response.status === 429) {
+      this.retryCount++
+      this.lastError = `[${response.status}] ${response.statusText}`
+      if (response.headers) {
+        const rateLimitResetTimestamp = parseInt(
+          response.headers['X-RateLimit-Reset'],
+          10
+        )
+        if (isFinite(rateLimitResetTimestamp)) {
+          timeUntilRefreshInMs =
+            rateLimitResetTimestamp -
+            Date.now() +
+            this.clockSkewInSeconds * 1000
+        } else {
+          timeUntilRefreshInMs = 5 * 1000
+        }
+        // We want subsequent calls to get_token to be able to interrupt our
+        //  Timeout when it's waiting for e.g. a long normal expiration, but
+        //  not when we're waiting for a rate limit reset. Sleep instead.
+        await sleep(timeUntilRefreshInMs)
+        timeUntilRefreshInMs = 0
+      }
+    } else if ([400, 401, 415].includes(response.status)) {
+      // Unrecoverable errors
+      this.retryCount = 0
+      this.tokenEmitter.emit('access_token', {
+        error: new Error(`[${response.status}] ${response.statusText}`),
+      })
+      this.stopPoller()
+      return
+    } else {
+      this.retryCount++
+      this.lastError = new Error(`[${response.status}] ${response.statusText}`)
+      timeUntilRefreshInMs = backoff({
+        attempt: this.retryCount,
+        minTimeout: 25,
+        maxTimeout: 1000,
+      })
+    }
+
+    if (this.retryCount % this.maxRetries == 0) {
+      this.tokenEmitter.emit('access_token', { error: this.lastError })
+      // TODO: figure out timing and whether to reset retries?
+    }
+    this.pollerTimer = setTimeout(
+      this.pollerLoop.bind(this),
+      timeUntilRefreshInMs
+    ).unref()
```

Comment:
I think we should update all the `setTimeout().unref()` calls to `setTimeout()?.unref()`, that way `unref()` is only called if it is defined. I believe it is not defined in cloudflare (we use this pattern elsewhere because of that)